### PR TITLE
Abort parsing when invalid clause identifiers are seen

### LINKED ISSUE #232: Odd error message when parsing not valid policy condition
### Before opening, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/cedar-policy/cedar/issues?q=is%3Aissue+).
- [X] I have read the guide for [submitting bug reports](https://github.com/cedar-policy/cedar/blob/main/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### Bug Category

Cedar Parser

### Describe the bug

Odd error message when trying to parse this policy

```
permit (principal, action, resource)
whenless {};
```
>  poorly formed: not a valid policy condition: whenless
>  poorly formed: whenless clause should not be empty

If `whenless` is not valid, then why say it should not be empty?

### Expected behavior

A single error message would be best (preferably with line/character offsets)

> poorly formed: not a valid policy condition: whenless

### Reproduction steps

Save the above policy into a file named `whenless.cedar` then use the `cedar` cli tool

```
cedar check-parse -p whenless.cedar 
failed to parse policy set from file whenless.cedar: 
  poorly formed: not a valid policy condition: whenless
  poorly formed: whenless clause should not be empty
```

### Code Snippet

_No response_

### Log output

_No response_

### Additional configuration

_No response_

### Operating  System

_No response_

### Additional information and screenshots

_No response_

### PR DESCRIPTION
**Pull request by shaobo-he-aws**
Fixes #232

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [ ] A breaking change requiring a major version bump to `cedar-policy` (e.g., changes to the signature of an existing API).
- [ ] A backwards-compatible change requiring a minor version bump to `cedar-policy` (e.g., addition of a new API).
- [ ] A bug fix or other functionality change requiring a patch to `cedar-policy`.
- [x] A change "invisible" to users (e.g., documentation, changes to "internal" crates like `cedar-policy-core`, `cedar-validator`, etc.)
- [ ] A change (breaking or otherwise) that only impacts unreleased or experimental code.

I confirm that this PR (choose one, and delete the other options):

- [ ] Updates the "Unreleased" section of the CHANGELOG with a description of my change (required for major/minor version bumps).
- [x] Does not update the CHANGELOG because my change does not significantly impact released code.

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [x] Does not require updates because my change does not impact the Cedar Dafny model or DRT infrastructure.
- [ ] Requires updates, and I have made / will make these updates myself. (Please include in your description a timeline or link to the relevant PR in `cedar-spec`, and how you have tested that your updates are correct.)
- [ ] Requires updates, but I do not plan to make them in the near future. (Make sure that your changes are hidden behind a feature flag to mark them as experimental.)
- [ ] I'm not sure how my change impacts `cedar-spec`. (Post your PR anyways, and we'll discuss in the comments.)

## Disclaimer

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: cedar-policy-core/src/parser/cst_to_ast.rs [hunk 1]
```diff
@@ -650,11 +650,13 @@ impl ASTNode<Option<cst::Cond>> {
...
-            (Some(false), Some(e)) => Some(construct_expr_not(e, src.clone())),
-            _ => None,
-        }
+        maybe_expr.map(|e| {
+            if maybe_is_when {
+                e
+            } else {
+                construct_expr_not(e, src.clone())
+            }
+        })
```
Comment (khieta) [inline cedar-policy-core/src/parser/cst_to_ast.rs:659]: I think this change results in a different program logic -- what happens when `maybe_expr` is `None`? The quick fix would be to add a question mark on line 651.

OTOH, given that the CI passed, perhaps this logic wasn't important. Or we're just not doing a good job of testing parse errors :)

**Comment by john-h-kastner-aws** [comment]
The error `poorly formed: whenless clause should not be empty` is odd, but still somewhat useful, so I'm not convinced that eliminating it entirely for a typoed keyword is best. Could we instead check if the keyword not `when` or `unless`, and then generate a different, more generic, error message like `poorly formed: policy body clause should not be empty`?

**Comment by cdisselkoen** [comment]
I second John's suggestion, but "policy body clause" might not be the clearest wording, maybe "policy clause"?  What do our docs call these?

File: cedar-policy-core/src/parser/cst_to_ast.rs (hunk 1 above)
Comment (cdisselkoen) [reply cedar-policy-core/src/parser/cst_to_ast.rs:659]: On my reading, the behavior when `maybe_expr` is `None` is unchanged -- `Option::map` returns `None` if the receiver is `None`

Comment (khieta) [reply cedar-policy-core/src/parser/cst_to_ast.rs:659]: Haha good point. I missed the `map`.

**Comment by khieta** [comment]
`when {...}` and `unless {...}` are "conditions", so I think the `{...}` would be the "condition body".

But I feel that the pair of errors
```
  poorly formed: not a valid policy condition: whenless
  poorly formed: condition body should not be empty
```
are still confusing together. I'd still prefer just the first (possibly modified to point out that the valid options are "when" and "unless"). Then after the user fixes the first error, they would get the second.

**Comment by cdisselkoen** [comment]
That makes sense to me too

**Comment by cdisselkoen** [comment]
As a meta point, I don't think that "poorly formed" adds anything to these messages (or other messages) -- should we remove it from all `ToAST` parse errors?

**Comment by shaobo-he-aws** [comment]
> As a meta point, I don't think that "poorly formed" adds anything to these messages (or other messages) -- should we remove it from all `ToAST` parse errors?

Andrew is working on ToAST errors. We can bring the suggestion in his PR.

**Comment by shaobo-he-aws** [comment]
> The error `poorly formed: whenless clause should not be empty` is odd, but still somewhat useful, so I'm not convinced that eliminating it entirely for a typoed keyword is best. Could we instead check if the keyword not `when` or `unless`, and then generate a different, more generic, error message like `poorly formed: policy body clause should not be empty`?

I think this PR won't remove the error message `... clause should not be empty`. Like @khieta pointed out, it would show up after the user corrects the identifier.

**Comment by john-h-kastner-aws** [comment]
> it would show up after the user corrects the identifier.

I don't agree that this is better than showing up at the same time. We know that's the body shouldn't be empty, regardless of what type of body it's meant to be.

**Review by khieta — APPROVED** [review]
I'm happy with the proposed change. I prefer the current implementation, but would also be ok with @john-h-kastner-aws's alternative.

**Review by cdisselkoen — APPROVED** [review]
After consideration, I'm on the same page as Kesha's comment immediately above.

**Pull request merged**